dali/python/nvidia/dali/plugin/pytorch/experimental/proxy/__init__.py

```diff
@@ -0,0 +1,517 @@
+                if err is not None:
+                    raise err
+
+                if curr_batch_id == req_batch_id:
+                    req_outputs = curr_processed_outputs
+                else:
+                    self.cache_outputs[curr_batch_id] = curr_processed_outputs
+        # Unpack single element tuples
+        if isinstance(req_outputs, tuple) and len(req_outputs) == 1:
+            req_outputs = req_outputs[0]
+        return req_outputs
+
+    def produce_data(self, obj):
+        """
+        A generic function to recursively visits all elements in a nested structure and replace
+        instances of DALIPipelineRunRef with the actual data provided by the DALI server
+
+        Args:
+            obj: The object to map (can be an instance of any class).
+
+        Returns:
+            A new object where any instance of DALIPipelineRunRef has been replaced with actual
+            data.
+        """
+        # If it is an instance of DALIPipelineRunRef, replace it with data
+        if isinstance(obj, DALIPipelineRunRef):
+            return self._get_outputs(obj)
+        # If it is a custom class, recursively call produce data on its members
+        elif hasattr(obj, "__dict__"):
+            for attr_name, attr_value in obj.__dict__.items():
+                setattr(obj, attr_name, self.produce_data(attr_value))
+            return obj
+
+        # If it's a list, recursively apply the function to each element
+        elif isinstance(obj, list):
+            return [self.produce_data(item) for item in obj]
+
+        # If it's a tuple, recursively apply the function to each element (and preserve tuple type)
+        elif isinstance(obj, tuple):
+            return tuple(self.produce_data(item) for item in obj)
+
+        # If it's a dictionary, apply the function to both keys and values
+        elif isinstance(obj, dict):
+            return {key: self.produce_data(value) for key, value in obj.items()}
+
+        else:  # return directly anything else
+            return obj
+
+    def _thread_fn(self):
+        """
+        Asynchronous DALI thread that gets iteration data from the queue and schedules it
+        for execution
+        """
+        self.pipe.build()  # just in case
+
+        while not self.thread_stop_event.is_set():
+            try:
+                torch.cuda.nvtx.range_push("dali_input_q.get")
+                batch_id, inputs = self.dali_input_q.get(timeout=5)
+                torch.cuda.nvtx.range_pop()
+            except mp.TimeoutError:
+                continue
+            except Empty:
+                continue
+
+            err = None
+            torch_outputs = None
+            torch.cuda.nvtx.range_push(f"schedule iteration {batch_id}")
+            try:
+                for idx, input_name in enumerate(self.dali_input_names):
+                    self.pipe.feed_input(input_name, inputs[idx])
+                self.pipe._run_once()
+                pipe_outputs = self.pipe.outputs()
```

Comment:
Is there any way we can schedule more batches before we try to access them? Or will we always stick with processing one batch at a time?

Reply:
The problem is that otherwise we'd have to do `pipe.outputs` in a different thread, which is generally not safe. We can make it safe in the native code, but I would advise against it in the scope of this already very large PR.